### PR TITLE
feat(#5000): populate Bolt ResultSummary counters for write queries

### LINKED ISSUE #5000: Bolt: populate ResultSummary counters for write queries
Part of #4890. Epic #4882.

Covers conformance scenario **RESULT-004** (`bolt/conformance/spec.yaml`), currently `current_status: expected-fail`. Discovered while implementing #4885's Python conformance suite.

### Motivation
`ResultSummary` counters (`nodesCreated`, `relationshipsCreated`, `propertiesSet`, etc.) are **never populated** for Bolt write queries. Two layers are missing:
1. `BoltNetworkExecutor`'s `handleRun`/`handlePull`/`handleDiscard` build **no `stats` map** in the SUCCESS response metadata.
2. The underlying query engine has **no CRUD-counter tracking** to source those values from - Cypher's `CreateStep`/`SetStep`/`DeleteStep` don't aggregate counts, so even a Bolt-layer fix would have nothing to report.

This is the largest of the #4890 gaps because it is cross-cutting into the query engine, not writer-only.

### Proposed design
- Add lightweight CRUD-counter aggregation to the Cypher mutation steps (Create/Set/Delete/Merge/Remove), surfaced on the result/execution context. Keep it allocation-light and off the read hot path.
- Emit the counters as the Bolt `stats` map in the RUN/PULL SUCCESS metadata (`nodes-created`, `relationships-created`, `properties-set`, `nodes-deleted`, `relationships-deleted`, `labels-added`, etc.).
- Consider whether HTTP/SQL result summaries should surface the same counters (possible shared win; scope carefully to avoid overreach).

### Acceptance criteria
- [ ] RESULT-004 passes across all five per-language suites; `summary.counters()` reflects actual writes.
- [ ] Counter tracking adds no measurable overhead to read queries.
- [ ] No regression in existing write-query behavior.

### Dependencies
Query-engine change (Cypher mutation steps) + Bolt metadata wiring. Larger blast radius than the other #4890 children.

### Effort estimate
**M-L**

### PR DESCRIPTION
**Pull request by robfrank**
Closes #5000. Part of #4890 (Group C protocol/type-fidelity gaps), epic #4882 (Bolt Driver Compatibility Certification).

## Problem
`ResultSummary` counters (`nodesCreated`, `relationshipsCreated`, `propertiesSet`, ...) were never populated for Bolt write queries. Two layers were missing: (1) `BoltNetworkExecutor` built no `stats` map in the SUCCESS metadata, and (2) the native Cypher engine had no CRUD-counter tracking to source it from.

## Approach
A reusable, allocation-light engine capability, consumed by the Bolt layer:

- **`QueryStatistics`** - a mutable accumulator of primitive `int` counters (nodes/relationships created/deleted, properties-set, labels added/removed, indexes/constraints added/removed) with `containsUpdates()`.
- **`CommandContext.getStatistics()`** - lazily allocated, so pure read queries (`MATCH`/`RETURN`) never create one and take no new hot-path branch.
- **Mutation steps** (`CreateStep`, `SetStep`, `DeleteStep`, `MergeStep`, `RemoveStep`) increment the accumulator with Neo4j-compatible semantics; the DDL path (`OpenCypherQueryEngine.executeDDL`) counts index/constraint add/remove, guarded so `IF [NOT] EXISTS` no-ops don't over-count.
- **`ResultSet.getStatistics()`** - a new `default` method mirroring `getExecutionPlan()` (zero blast radius on other implementations); `CypherExecutionPlan.execute()` and `executeDDL()` attach the accumulator to the returned result set.
- **`BoltNetworkExecutor`** - `handlePull`/`handleDiscard` read it and emit a Neo4j hyphenated `stats` map (non-zero counters + `contains-updates`) on the terminal SUCCESS, only for writes.

Counters reflect the statement's work regardless of a later transaction rollback (Neo4j-correct).

## Conformance
Flips `RESULT-004` in `bolt/conformance/spec.yaml` from `expected-fail` to `passing`, and unwraps the expected-fail harness in all five per-language e2e suites (Java, JS, Python, Go, C#) so each asserts the real counters.

## Testing
- Engine unit tests (`QueryStatisticsTest`, `ResultSetStatisticsTest`, `CypherQueryStatisticsTest`): 21 cases covering CREATE/SET/DELETE/MERGE/REMOVE, `SET`/`REMOVE` label, `CREATE`/`DROP INDEX`/`CONSTRAINT`, distinct-label dedup, `SET n = {map}` replace semantics, absent-property removal (counts 0), and a read-query-allocates-nothing regression.
- Bolt IT (`Bolt5000ResultCountersIT`) via the real `neo4j-java-driver`: asserts `summary.counters()` for a write and no updates for a read.
- Cypher mutation + DDL regression subsets and the full `com.arcadedb.query.opencypher` package: green.
- Spec validator (`bolt/conformance/validate_spec.py`): 39 scenarios OK.

## Scope / follow-up
Scoped to the engine accumulator + Bolt consumption. Surfacing the same counters over HTTP `/command` and gRPC, aggregating counts across top-level `UNION` writes and `CALL { ... }` write subqueries (currently documented in-code as returning zero to the caller), and constraint-kind test coverage are tracked in #5015.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: engine/src/main/java/com/arcadedb/query/opencypher/query/OpenCypherQueryEngine.java [hunk 1]
```diff
@@ -304,26 +305,40 @@ private ResultSet execute(final String queryString, final CypherStatement statem
...
+  /**
+   * Checks whether a type index already covers exactly {@code propertyNames}, looking both at the
+   * type's own indexes and any inherited from a parent type. Used by Cypher constraint/index DDL to
+   * tell whether an {@code IF NOT EXISTS} create is a genuine no-op, so the DDL result statistics
+   * (indexes/constraints added) only count schema changes that actually happened.
+   */
+  private static boolean indexExistsOnProperties(final Schema schema, final String typeName, final String[] propertyNames) {
+    final DocumentType type = schema.getType(typeName);
+    return type.getIndexByProperties(propertyNames) != null || type.getPolymorphicIndexByProperties(propertyNames) != null;
+  }
```
[the remark this answers is not in the document]
Comment (robfrank) [reply engine/src/main/java/com/arcadedb/query/opencypher/query/OpenCypherQueryEngine.java:343]: Guarded with an `existsType` check in 46fef7de7. Two clarifications on the mechanism: in this codebase `LocalSchema.getType(...)` throws `SchemaException` on a missing type rather than returning `null`, so a `type == null` check would be dead code; and both callers (`executeCreateIndex`, `executeCreateConstraint`) auto-create the type via `getOrCreateVertexType/EdgeType` before this helper runs, so the path was not reachable. The `existsType` guard hardens it against a future caller regardless - a missing type now correctly reports "no index".

File: engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/SetStep.java [hunk 2]
```diff
@@ -296,6 +318,9 @@ private void applyMergeMap(final SetClause.SetItem item, final Result result,
     }
 
     mutableDoc.save();
+    // Every map entry mutates a property, whether set (non-null) or removed (null value).
+    final QueryStatistics stats = context.getStatistics();
+    stats.addPropertiesSet(map.size());
```
[the remark this answers is not in the document]
Comment (robfrank) [reply engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/SetStep.java]: Correct - fixed in 46fef7de7. The `+=` branch now counts each non-null assignment and each null-removal only when the property actually existed (`mutableDoc.has(...)` before `remove`), matching `applyReplaceMap`. Added a regression test (`SET n += {c:3, b:null, z:null}` on a node with `b` but not `z` asserts propertiesSet == 2).

File: engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/MergeStep.java [hunk 3]
```diff
@@ -1354,6 +1384,8 @@ else if (mapValue instanceof Document srcDoc)
             else
               mutableDoc.set(entry.getKey(), TemporalUtil.toCoreJavaType(entry.getValue()));
           mutableDoc.save();
+          // Every map entry mutates a property, whether set (non-null) or removed (null value).
+          context.getStatistics().addPropertiesSet(map.size());
```
[the remark this answers is not in the document]
Comment (robfrank) [reply engine/src/main/java/com/arcadedb/query/opencypher/executor/steps/MergeStep.java]: Same fix applied to the MERGE_MAP branch in 46fef7de7, with a regression test through `MERGE ... ON MATCH SET n += {a:null, absent:null, b:2}` asserting propertiesSet == 2 (removal of the absent key is not counted).

**Comment by robfrank** [comment]
Responses to the review, by item:

**1. UNION / CALL `contains-updates: false` (actively wrong).** Agreed this is the sharpest of the gaps: `type: "w"` but no `stats` map means a mutating statement reports `containsUpdates() == false`, which is wrong rather than merely incomplete. I considered forcing `contains-updates: true` for known-writes with zero granular counters, but that produces an internally inconsistent summary (updates happened, yet every counter is 0) and still misreports the actual work. The correct fix is propagating the child-context counters up through `executeUnion()`/`executeWithSeedRow()`. I have expanded follow-up #5015 to call this out explicitly as active incorrectness (not just missing counts), so it is tracked rather than silently shipped.

**2. `propertiesSet` from `getPropertyNames().size()` on CREATE.** Added a comment at the create sites noting the invariant that a freshly created record`\s property names are exactly the user-set properties (labels/type are not properties), so the coupling to the serializer surface is documented (46fef7de7).

**3. DETACH DELETE double-count.** Real edge case - `deleteAllEdges()`/`deleteVertex()` do not consult the batch `deleted` set, so a query that both `DETACH DELETE`s a node and explicitly deletes one of its edges could count that edge twice where Neo4j dedups. Not a regression and niche; added it to #5015 with a note to add the repro test there.

**4. Thread-safety.** Added a class Javadoc line to `QueryStatistics`: not thread-safe, single writer per command context, plain ints with no synchronization (46fef7de7). Correct that it is safe today because mutation steps run single-threaded per query; the note guards against a future parallel-mutation step.

**5. `putIfNonZero` long widening.** Intentional - Bolt integers are 64-bit; confirmed.

**6. `docs/superpowers/` spec/plan in-tree.** This is the established repo convention (spec/plan docs are committed alongside feature PRs here), so keeping them. Point taken on drift.

Also addressed Gemini's inline findings in 46fef7de7 (the `SET n += {map}` null-for-absent over-count in both `SetStep` and `MergeStep`, plus a defensive `existsType` guard on the DDL type lookup). Two new regression tests; `CypherQueryStatisticsTest` 21/21 and the mutation/DDL regression subset 170/170 green.

**Comment by robfrank** [comment]
Second-pass responses:

**1. Null-inline property on CREATE.** Verified this is already Neo4j-correct: `CreateStep.setProperties` skips null values (`// In Cypher, null property values are not stored; if (value != null) document.set(...)`), so `CREATE (n:Foo {a:1, b:null})` stores only `a` and `getPropertyNames().size() == 1`. Added a lock-down test asserting `propertiesSet == 1` for that shape (fdb6940b1). The auto-populated/default-property case does not arise for Cypher-created types (Cypher auto-creates types with no declared defaults); the coupling is documented at the create sites.

**2. UNION / CALL conformance hole.** Confirmed no conformance scenario in `spec.yaml` exercises a UNION or CALL write, so flipping RESULT-004 to passing does not certify a shape that misreports. On the "zero-but-wrong vs not-emitting" point: the current behavior is *not* emitting a `stats` map at all for these shapes (the guard is `containsUpdates()`, which is false when the outer context has no counts), so the driver reports `containsUpdates() == false` with no counters - there is no zero-valued map being sent. #5015 tracks propagating the real counts.

**3. Parent-context delegation in `getStatistics()`.** Good lead - noted in #5015 as a candidate mechanism. It would help the `CALL {}` case if `executeWithSeedRow` sets the outer context as parent, but UNION branches run as independent `execute()` calls with fresh root contexts, so they would still need explicit aggregation. Captured both in the follow-up rather than half-solving here.

**4. Replace-map rule duplicated.** After the `+=` fix in 46fef7de7 the `!prop.startsWith("@") && map.get(prop) == null` idiom remains in the two REPLACE_MAP branches (`SetStep.applyReplaceMap`, `MergeStep`); single-sourcing it into a small helper is a clean maintainability win. Added to #5015 to avoid a refactor mid-review-cycle.

**5. Thread-safety Javadoc.** Added in 46fef7de7: "Not thread-safe: a single command context is written by one query thread at a time; counters are plain ints with no synchronization." Confirmed no write path dispatches mutation steps onto the async/parallel pools.

**6. `containsUpdates()` incl. index/constraint.** Intentional, matching Neo4j `SummaryCounters.containsUpdates()`; `containsSystemUpdates()` is deliberately not modeled (design non-goal).

**Comment by robfrank** [comment]
Third-pass responses:

**1. DETACH DELETE relationship undercount - confirmed and fixed (033332246).** You are right: the immediate path (`deleteVertex` -> `deleteAllEdges`) counts cascaded edges, but the deferred/FOREACH path (`flushDeferredDeletes`, used when `DEFERRED_DELETE_BATCH_VAR` is set) called `v.delete()` without counting them, so `FOREACH (x IN xs | DETACH DELETE x)` on a connected node reported `relationshipsDeleted: 0`. Fixed by counting each connected edge via the existing `deleteObjectStatic` before `v.delete()` in the detach branch - which also respects the `deleted` set, so an edge explicitly deleted earlier in the same batch is not double-counted (closing the double-count concern from the first pass in the same spot). Added two tests: plain `MATCH ... DETACH DELETE` (immediate path, was already correct) and `FOREACH ... DETACH DELETE` (deferred path, was red at relationshipsDeleted 0, now 1).

**2. SET vs MERGE ON MATCH SET on unchanged values - aligned (033332246).** The MergeStep PROPERTY branch counted inside the `valuesEqual` MVCC guard, so `MERGE (n) ON MATCH SET n.p = <same>` reported `properties-set: 0` while `SetStep` counts unconditionally. Neo4j counts the assignment regardless of whether the value changed, so I moved the `addPropertiesSet(1)` out of the guard - the counter now matches Neo4j and SetStep, while the write-skip optimization (and its ConcurrentModificationException avoidance) is preserved. Added a regression test.

**3. `handleDiscard` read-before-drain.** Added a one-line comment at the read site reaffirming the invariant: statistics are computed eagerly when the write is materialized in the query plan, so they are valid to read before draining.

**4. Docs footprint.** Intentional and per repo convention (spec/plan docs are committed with feature PRs here).

`CypherQueryStatisticsTest` now 25/25; Delete/Foreach/Merge/Set regression 121/121 green.

**Comment by robfrank** [comment]
Fourth-pass responses:

**1. Retry double-count in CreateStep - confirmed and fixed (7886b2901).** Verified: `CreateStep.createPatterns`/`createPatternsBatch` wrap creation in `database.transaction(block, true)`, which retries on MVCC/duplicate-key conflict, and the counter increments run inside the block, so a standalone `CREATE` that actually retries inflated the counts. Fixed by snapshotting the counters before the transaction (`QueryStatistics.copy()`) and restoring them at the top of the transaction lambda (`stats.restore(snapshot)`), so each re-attempt starts from the pre-attempt baseline and only the successful attempt`\s deltas survive. Also clears the batch `createdResults` list at the top of each attempt (a retry would otherwise duplicate result rows). Checked the other steps as you suggested: `MergeStep`, `SetStep`, and `DeleteStep` all use manual `begin()/commit()` with no retry loop, so they are not affected - no change there. Added a `copy()/restore()` unit test; deterministically forcing a `database.transaction` retry needs a concurrent MVCC/duplicate-key conflict (impractical as a stable unit test), so the mechanism is covered by that test plus the reset-at-attempt-start placement.

**2. CALL / UNION zero counters + parent-walk suggestion.** Deferred to #5015, where I already captured exactly your suggestion: have `BasicCommandContext.getStatistics()` walk to the root context (as `getDatabase()` does). Noted there that it closes CALL if the child sets the outer as parent, but UNION branches run as independent top-level `execute()` calls and still need per-branch aggregation.

**3. Constraint "existed before" edge case (plain index present, then UNIQUE/KEY added -> constraints-added not incremented).** Real edge case in the counter semantics; added it to #5015 alongside the NOT_NULL/KEY/TYPED coverage gap.

**Nit (docs in-tree).** Intentional, per repo convention for feature PRs.

`QueryStatisticsTest` 4/4 (incl. the new copy/restore case), `CypherQueryStatisticsTest` 25/25, Create/Merge regression green.

**Comment by robfrank** [comment]
Fifth-pass responses:

**1. Self-loop DETACH DELETE double-count - confirmed and fixed (8b26b0b40).** Good catch. `deleteAllEdges` (the immediate path) collected a self-loop from both OUT and IN with no dedup, and a new test reproduced it: `MATCH (a:SL) DETACH DELETE a` on `(a)-[:LOOP]->(a)` reported `relationshipsDeleted: 2`. Fixed by de-duplicating collected edges by identity (a `Set<RID>`) so a self-loop is counted once, and wrapping `edge.delete()` in a `RecordNotFoundException` guard so an already-removed edge (e.g. a shared edge deleted when the other endpoint was detached) is never counted twice or throws - matching the robustness of the deferred path. Test now asserts `relationshipsDeleted: 1`.

**Nit (contains-updates).** Added a note at `BoltResultStats.toStatsMap` that `contains-updates` is emitted unconditionally, so callers must only invoke it when `containsUpdates()` is true (8b26b0b40).

**2. Read-only `getStatistics()` accessor.** No current read-side caller allocates through it (`execute()` only calls it on the write branch), so this is a future footgun rather than a live issue; adding a peek-only accessor now would be speculative. Noted for the #5015 follow-up if a read-side consumer (HTTP/gRPC) needs it.

**Nit (Bolt read-after-close helper) + item 3 (QueryStatistics package).** The `handleDiscard` read site documents the eager-materialization invariant; a shared read-after-close helper is a reasonable robustness refactor but only matters if a lazy write path is ever introduced (none exists), so I have left it. `QueryStatistics` sits with `CommandContext`/`ResultSet` in `com.arcadedb.query.sql.executor`, which is the shared executor home both the Cypher engine and Bolt already depend on - intentional.

`CypherQueryStatisticsTest` 26/26, `OpenCypherDeleteTest` 9/9 green.

**Comment by robfrank** [comment]
Sixth-pass responses:

**1. `copy()` dropped statistics - fixed (latest commit).** `BasicCommandContext.copy()` propagated variables/parent/child/profiling but not the accumulator, so a step running against a `context.copy()` would have its increments silently vanish. Fixed by sharing the same `QueryStatistics` reference in `copy()`, so mutations through a copied context aggregate into one place. Added a unit test asserting `copy().getStatistics()` is the same instance and increments through the copy land on the original total. The broader hardening you suggested - `getStatistics()` walking to the root parent (mirroring `getVariable`) - is the natural mechanism to close the UNION/CALL gap and is tracked in #5015; I kept that out of this PR since it changes UNION/CALL behavior, but the `copy()` fix removes the silent-vanish trap now.

**2. MERGE retry over-count - confirmed safe, no guard needed.** Verified: `MergeStep` opens its transaction with manual `begin()`/`commit()` (no `database.transaction(..., true)` retry wrapper and no retry loop), and neither `OpenCypherQueryEngine` nor `CypherExecutionPlan` wraps step execution in a retrying transaction - only `CreateStep` uses `database.transaction(..., true)`, which is why only it carries the snapshot/restore guard. A MERGE create block is never re-executed by a retry, so it cannot over-count; a mid-commit conflict fails the statement rather than re-running the body.

**3a. Retry/restore test.** Already present as `QueryStatisticsTest.copyAndRestoreRollBackIncrements` (added with the CreateStep retry fix): it snapshots, re-increments, restores, and asserts the counters roll back - exactly the contract you describe.

**3b. Constraint-kind counting tests (NOT NULL / KEY / TYPED, and the no-op `IF NOT EXISTS` guards).** Tracked in #5015 alongside the plain-index-then-constraint `existedBefore` edge case.

`QueryStatisticsTest` 5/5 green.
